sys/vane/alef.hoon

```diff
@@ -0,0 +1,2588 @@
+    ::  load mutant environment
+    ::
+    =.  packet-pump  core.acc
+    ::  if we can't send any more packets, we're done
+    ::
+    ?:  =(0 num-slots.acc)
+      [`val stop=%.y acc]
+    ::  if the packet hasn't expired, we're done
+    ::
+    ?:  (gte expiry.val now.channel)
+      [`val stop=%.y acc]
+    ::  packet has expired so re-send it
+    ::
+    =/  =static-fragment
+      =>  [key val]
+      [message-num num-fragments fragment-num fragment]
+    ::
+    =.  packet-pump    (give %send static-fragment)
+    =.  metrics.state  (on-resent:gauge -.val)
+    ::  update $sent-packet-state in .val and continue
+    ::
+    =.  expiry.val     (next-retry-expiry:gauge -.val)
+    =.  sent-date.val  now.channel
+    =.  retried.val    %.y
+    ::
+    [`val stop=%.n (dec num-slots.acc) packet-pump]
+  ::  +feed: try to send a list of packets, returning unsent and effects
+  ::
+  ++  feed
+    |=  fragments=(list static-fragment)
+    ^+  [fragments gifts state]
+    ::  return unsent back to caller and reverse effects to finalize
+    ::
+    =-  [unsent (flop gifts) state]
+    ::
+    ^+  [unsent=fragments packet-pump]
+    ::  resend lost packets first, possibly adjusting congestion control
+    ::
+    =.  packet-pump  resend-lost
+    ::  bite off as many fragments as we can send
+    ::
+    =/  num-slots  num-slots:gauge
+    =/  sent       (scag num-slots fragments)
+    =/  unsent     (slag num-slots fragments)
+    ::
+    :-  unsent
+    ^+  packet-pump
+    ::  if nothing to send, we're done
+    ::
+    ?~  sent  packet-pump
+    ::  convert $static-fragment's into +ordered-set [key val] pairs
+    ::
+    =/  send-list
+      %+  turn  sent
+      |=  static-fragment
+      ^-  [key=live-packet-key val=live-packet-val]
+      ::
+      :-  [message-num fragment-num]
+      :-  :+  expiry=next-expiry:gauge
+            sent-date=now.channel
+          retried=%.n
+      [num-fragments fragment]
+    ::  update .live and .metrics
+    ::
+    =.  live.state     (gas:packet-queue live.state send-list)
+    =.  metrics.state  (on-sent:gauge (lent send-list))
+    ::  TMI
+    ::
+    =>  .(sent `(list static-fragment)`sent)
+    ::  emit a $packet-pump-gift for each packet to send
+    ::
+    |-  ^+  packet-pump
+    ?~  sent  packet-pump
+    =.  packet-pump  (give %send i.sent)
+    $(sent t.sent)
+  ::  +on-hear: handle ack on a live packet
+  ::
+  ::    Traverse .live from the head, marking packets as lost until we
+  ::    find the acked packet. Then delete the acked packet and try to
+  ::    resend lost packets.
+  ::
+  ::    If we don't find the acked packet, no-op: no mutations, effects,
+  ::    or resending of lost packets.
+  ::
+  ++  on-hear
+    |=  [=message-num =fragment-num]
+    ^+  packet-pump
+    ::
+    =-  ::  if no sent packet matches the ack, don't apply mutations or effects
+        ::
+        ?.  found.-
+          packet-pump
+        ::
+        =.  metrics.state  metrics.-
+        =.  live
```

Comment:
Any particular reason not to use `+sign:as:crub` and `+sure:as:crub` here?

Reply:
I wanted more manual control while I was designing this. In particular, the order in which we validate life, public key, and signature could affect DDoS susceptibility. So I decided to do these things more manually until they settle.

I wouldn't be surprised if we go back to +sign and +sure after thinking a bit harder.